hal-x86_64/src/vga.rs

```diff
@@ -0,0 +1,175 @@
+use core::{fmt, ptr};
+use mycelium_util::sync::spin;
+
+lazy_static::lazy_static! {
+    static ref BUFFER: spin::Mutex<Buffer> = spin::Mutex::new(Buffer {
+        col: 0,
+        row: 0,
+        color: ColorSpec::new(Color::Green, Color::Black),
+        buf: unsafe { &mut *(0xb8000 as *mut Buf) },
+    });
+}
+
+pub fn writer() -> Writer {
+    Writer(())
+}
+
+#[derive(Debug)]
+pub struct Writer(());
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(u8)]
+pub enum Color {
+    Black = 0,
+    Blue = 1,
+    Green = 2,
+    Cyan = 3,
+    Red = 4,
+    Magenta = 5,
+    Brown = 6,
+    LightGray = 7,
+    DarkGray = 8,
+    LightBlue = 9,
+    LightGreen = 10,
+    LightCyan = 11,
+    LightRed = 12,
+    Pink = 13,
+    Yellow = 14,
+    White = 15,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct ColorSpec(u8);
+
+pub struct Buffer {
+    col: usize,
+    row: usize,
+    color: ColorSpec,
+    buf: &'static mut Buf,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(C)]
+struct Character {
+    ch: u8,
+    color: ColorSpec,
+}
+
+const BUF_HEIGHT: usize = 25;
+const BUF_WIDTH: usize = 80;
+
+type Buf = [[Character; BUF_WIDTH]; BUF_HEIGHT];
+
+impl ColorSpec {
+    pub const fn new(fg: Color, bg: Color) -> Self {
+        let bg = (bg as u8) << 4;
+        Self(bg | (fg as u8))
+    }
+}
+
+impl Character {
+    fn write(&mut self, ch: Character) {
+        unsafe {
+            // safety: we have mutable access to this character.
+            ptr::write_volatile(self as *mut Character, ch)
+        }
+    }
+}
+
+impl Buffer {
+    pub fn set_color(&mut self, color: ColorSpec) {
+        self.color = color;
+    }
+
+    fn blank(&self) -> Character {
+        self.character(b' ')
+    }
+
+    fn character(&self, ch: u8) -> Character {
+        Character {
+            ch,
+            color: self.color,
+        }
+    }
+
+    fn write(&mut self, s: &str) {
+        assert!(s.is_ascii(), "VGA buffer does not support Unicode");
+        for ch in s.bytes() {
+            self.write_char(ch);
+        }
+    }
+
+    fn write_char(&mut self, ch: u8) {
+        // TODO(eliza): assert char is ascii?
```

Comment:
the thing we're writing to (i believe?) speaks [CP437](https://en.wikipedia.org/wiki/Code_page_437), so any u8 is a valid character to write

Reply:
yeah, i guess if we're making assertions about ascii-ness, it should only be when trying to write a _string slice_ to the vga buffer...i was thinking more "avoid accidentally printing unicode"

Reply:
OH LOL we already do that when printing string slices, nvm.

Reply:
yeah that makes sense. i think checking that in `write(self, &str)` would be better so we can style on the vga buffer later (i almost want to volunteer a tui but i won't say it)

Reply:
of course nothing forbids explicit style functions here to do that

Reply:
> i think checking that in `write(self, &str)` would be better

yeah we already do that :)